lib/engine/game/g_1847_ae/step/draft.rb

```diff
@@ -0,0 +1,145 @@
+
+          def setup
+            @companies = @game.companies.select { |c| c.owner.nil? && !c.closed? }
+            @companies = @companies.sort_by { |item| [item.revenue, item.value] }
+          end
+
+          def available
+            @companies
+          end
+
+          def active?
+            true
+          end
+
+          def may_purchase?(_company)
+            true
+          end
+
+          def auctioning; end
+
+          def bids
+            {}
+          end
+
+          def visible?
+            true
+          end
+
+          def players_visible?
+            true
+          end
+
+          def tiered_auction_companies
+            @companies.group_by(&:revenue).values
+          end
+
+          def name
+            'Draft'
+          end
+
+          def description
+            'Draft Private Companies'
+          end
+
+          def finished?
+            @game.draft_finished = @companies.empty?
+
+            @companies.empty? || entities.all?(&:passed?)
+          end
+
+          def actions(entity)
+            return [] if finished?
+
+            entity == current_entity ? ACTIONS : []
+          end
+
+          def process_bid(action, _suppress_log = false)
+            company = action.company
+            player = action.entity
+            price = action.price
+
+            company.owner = player
+            player.companies << company
+            player.spend(price, @game.bank)
+
+            @companies.delete(company)
+
+            @log << "#{player.name} buys #{company.name} for #{@game.format_currency(price)}"
+
+            company.abilities.each do |ability|
+              next unless ability.type == :shares
+
+              # Give related share to the player and give money for the share to the corporation
+              ability.shares.each do |share|
+                @game.share_pool.buy_shares(player, share, exchange: :free)
+                share.corporation.cash += share.corporation.par_price.price * share.percent / 10
+              end
+            end
+
+            # PLP company is only a temporary holder for the L presidency
+            company.close! if company.id == 'PLP'
+
+            entities.each(&:unpass!)
+            action_finalized
+            next_player! unless finished?
+          end
+
+          def process_pass(action, suppress_log = false)
+            @log << "#{action.entity.name} passes" unless suppress_log
+            action.entity.pass!
+            action_finalized
+            next_player! unless finished?
+          end
+
+          def pass_if_no_valid_action!
+            return if @companies.any? { |c| current_entity.cash >= min_bid(c) }
+
+            @log << "#{current_entity.name} has no valid actions and passes"
+            @round.process_action(Engine::Action::Pass.new(current_entity), suppress_log: true)
+          end
+
+          def next_player!
+            @round.next_entity_index!
+
+            pass_if_no_valid_action!
+          end
```

Comment:
This feels over complicated. Can't you put the `@companies_any?` check in `actions`, returning `[]` if the player can't afford anything. The log message would be sent from `log_skip`. This would mean you wouldn't need to have the almost unmodified version of `process_action` in the round class.

Reply:
That was the starting point since it's done that way in 18CZ, but I couldn't get that to work. First player for the next round sometimes wasn't properly set if players were skipped and if the first player in draft was to be skipped (which can happen here), the round was ending immediately.

Reply:
> That was the starting point since it's done that way in 18CZ, but I couldn't get that to work. First player for the next round sometimes wasn't properly set if players were skipped. If the first player in draft was to be skipped (which can happen here), the round was ending immediately.

Yes, I see this now. This is because the inherited version of `Round::Base.finished?` isn't finding any active steps, and so is ending the round. I think you're going to need to borrow some of the logic from `Round::Stock` to work out when to move to the next player and when to end the round. Let me think about this…

Reply:
Galatolol/18xx#1 is an attempt to move the code controlling the flow of the round out from the Step class into the Round class. See if that makes sense to you.

Reply:
Thank you. I'm looking into that.

Reply:
The proposed solution, with some tweaks, did the trick.